gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util.concurrent;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Iterables;
+
+import lombok.Synchronized;
+
+/**
+ * Base class for a task scheduler that can run {@link ScheduledTask}s periodically. Subclasses can implement
+ * {@link #startImpl(Optional)} and {@link #closeImpl()}.
+ */
+public abstract class TaskScheduler<K, T extends ScheduledTask<K>> implements Closeable {
+  private final Cache<K, CancellableTask<K, T>> cancellableTaskMap = CacheBuilder.newBuilder().build();
+  private volatile boolean isStarted = false;
+
+  protected TaskScheduler() {
+  }
+
+  @Synchronized
+  final void start(Optional<String> name) {
+    if (!this.isStarted) {
+      this.startImpl(name);
+      this.isStarted = true;
+    }
+  }
+
+  abstract void startImpl(Optional<String> name);
+
+  /**
+   * Schedules a subclass of {@link ScheduledTask} to run periodically.
+   *
+   * @param task the subclass of {@link ScheduledTask} to run every period
+   * @param period the period between successive executions of the task
+   * @param unit the time unit of the initialDelay and period parameters
+   */
+  public final void schedule(final T task, final long period, final TimeUnit unit) throws IOException {
+    Preconditions.checkArgument(isStarted, "TaskScheduler is not started");
+    try {
+      this.cancellableTaskMap.get(task.getKey(),
```

Comment:
If a task with that key is already present, this will silently not schedule the new task. I think it should throw an exception instead.